### PR TITLE
Android

### PR DESCRIPTION
**Pull request by sajacy**
The core Android compatibility problem is solved in the `www/js/compat.js` implementation.  There were also issues with exporting to CSV and some unguarded API calls.  

Another major issue was the build process.  I removed all build-generated files, but found that there were some customizations within the `lib/gauge.js` package, version resolution problems, porting issues with `cca`, and some general quirkiness in the setup.  

I forked the gauge.js project and patched it so we can use it in a more sane way (you'll find a bower reference to `sajacy-gauge.js`).  Also normalized the build to `npm install`, followed by `gulp` and `cca` as the only two actual build tools (`ionic` may not even be necessary anymore, but my dev environment is too dirty right now to test properly) Consequently, updated the README with build procedures.

Additionally, by default, the app actually installs into Chrome as a proper Chrome App, launchable outside Chrome.  This was a happy accident of trying to revamp the build process.  I've preserved the old behavior, which can be built via `gulp extension`.

### CODE REVIEW

**Comment by sajacy** [comment]
@georgiamoon @gembarrett @collina

**Comment by sajacy** [comment]
Forgot to mention.  Tested this on a Samsung S3, SGH-T999 variant, running Android 4.3 (jelly bean), non-rooted/OEM (AFAIK)

**Comment by georgiamoon** [comment]
So, I'm trying to `npm install -g cca`, and i'm getting the following errors:

npm WARN deprecated npmconf@2.1.2: this package has been reintegrated into npm and is now out of date with respect to npm
npm WARN deprecated graceful-fs@3.0.8: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0 as soon as possible.
npm WARN deprecated lodash@1.0.2: lodash@<3.0.0 is no longer maintained. Upgrade to lodash@^4.0.0.
npm WARN deprecated graceful-fs@1.2.3: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0 as soon as possible.
/usr/local/lib
└── (empty)

npm ERR! Darwin 14.5.0
npm ERR! argv "/usr/local/Cellar/node/5.6.0/bin/node" "/usr/local/bin/npm" "install" "-g" "cca"
npm ERR! node v5.6.0
npm ERR! npm  v3.6.0
npm ERR! path /usr/local/lib/node_modules/.staging/balanced-match-dc025c5f
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename '/usr/local/lib/node_modules/.staging/balanced-match-dc025c5f' -> '/usr/local/lib/node_modules/cca/node_modules/cordova-lib/node_modules/balanced-match'
npm ERR! enoent ENOENT: no such file or directory, rename '/usr/local/lib/node_modules/.staging/balanced-match-dc025c5f' -> '/usr/local/lib/node_modules/cca/node_modules/cordova-lib/node_modules/balanced-match'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/georgiabullen/code/measure-app-legacy/npm-debug.log
npm ERR! code 1

I've tried independently installing balanced-match, but that doesn't appear to help.

suggestions?

**Comment by sajacy** [comment]
Usually /usr/local/lib errors mean you need to `sudo npm install -g ...`

**Comment by georgiamoon** [comment]
Same error. I think it something with versions of node, but I'm not positive. Anyone else having this issue? @cda @critzo @natmey ?

**Comment by sajacy** [comment]
Hm, odd. You could try installing nvm first (Google for creationix nvm) which doesn't try to write to that directory. I tried with your version of node (5.6) via nvm and was able to finish successfully.

**Comment by georgiamoon** [comment]
OK -- nvm, and then using v5.0 worked to build. now I will test!

**Comment by georgiamoon** [comment]
I finally got it working!! Now to do real testing.

**Comment by georgiamoon** [comment]
Just for documentation purposes:

Issues:
- Need to add this step: 'cordova plugin add --save cordova-plugin-x-socialsharing' and then export works like sharing
- export still not quit working, @sajacy and I are troubleshooting
- Sometimes the app hangs on test start -- @sajacy -- is there a way to get logs from the device testing so we can see why this might be happening?

Enhancement thought:
- we might want to add a icon for the tray to allow a user to know that the app is running, similar to the icon feedback in the browser. Is this possible?

**Comment by georgiamoon** [comment]
Additional problem that I'm seeing: It seems that whenever the app detects a network connection change, it surfaces to the top on the device, and clobbers anything that the user is doing.

This also seems to result in a crash notification.

**Comment by sajacy** [comment]
Thanks @georgiamoon for the bug report and debug log.  I saw this error multiple times in the log:

```
02-19 11:20:40.078  4425 30389 I ActivityManager: START u0 {flg=0x10010004 cmp=org.measurementlab.apps.Measure/org.chromium.BackgroundActivity} from uid 10187 on display 0
02-19 11:20:40.153  4425  5694 D WifiService: acquireWifiLockLocked: WifiLock{NlpWifiLock type=2 binder=android.os.BinderProxy@5200ee0}
02-19 11:20:40.165  4425  5656 I chatty  : uid=1000(system) WifiScanningSer expire 6 lines
02-19 11:20:40.219  4425 24678 I ActivityManager: START u0 {cat=[android.intent.category.DEFAULT] flg=0x10004 cmp=org.measurementlab.apps.Measure/.MainActivity} from uid 10187 on display 0
02-19 11:20:40.228 16038 16038 W Activity: An activity without a UI must call finish() before onResume() completes
02-19 11:20:40.234 16038 16038 D AndroidRuntime: Shutting down VM
02-19 11:20:40.271 16038 16038 E AndroidRuntime: FATAL EXCEPTION: main
02-19 11:20:40.271 16038 16038 E AndroidRuntime: Process: org.measurementlab.apps.Measure, PID: 16038
02-19 11:20:40.271 16038 16038 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {org.measurementlab.apps.Measure/org.chromium.BackgroundActivity}: java.lang.IllegalStateException: Activity {org.measurementlab.apps.Measure/org.chromium.BackgroundActivity} did not call finish() prior to onResume() completing
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3103)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3134)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2481)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.ActivityThread.-wrap11(ActivityThread.java)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.os.Handler.dispatchMessage(Handler.java:102)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.os.Looper.loop(Looper.java:148)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.ActivityThread.main(ActivityThread.java:5417)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at java.lang.reflect.Method.invoke(Native Method)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
02-19 11:20:40.271 16038 16038 E AndroidRuntime: Caused by: java.lang.IllegalStateException: Activity {org.measurementlab.apps.Measure/org.chromium.BackgroundActivity} did not call finish() prior to onResume() completing
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.Activity.performResume(Activity.java:6339)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3092)
02-19 11:20:40.271 16038 16038 E AndroidRuntime:    ... 10 more
```

Which is a bug documented by this blog post:
https://commonsware.com/blog/2015/11/02/psa-android-6p0-theme.nodisplay-regression.html

It affects Android 6.0+ and apps that use background activities (such as this, via the [`cordova-plugin-background-app`](https://github.com/MobileChromeApps/cordova-plugin-background-app) plugin)

The log also indicates that the activity may have clobbered Slack a couple times (and maybe Twitter, Instagram, or Facebook too)...  can you confirm that - and maybe note if there were any notable circumstances - it seems ActivityManager "force quit" a bunch of apps (including the Measure app) - and then the Measure app came back to life via the alarm (which fires every minute) via the background-app plugin mentioned above.

Here's the snippet that shows the alarm firing, then creating the Measure window and pausing Slack:

```
02-19 11:20:39.788  4425  4445 I am_proc_start: [0,16038,10187,org.measurementlab.apps.Measure,broadcast,org.measurementlab.apps.Measure/org.chromium.ChromeAlarmsReceiver]
02-19 11:20:39.942  4425  7101 I am_proc_bound: [0,16038,org.measurementlab.apps.Measure]
02-19 11:20:40.078  4425 30389 I am_home_stack_moved: [0,0,1,1,startingNewTask]
02-19 11:20:40.078  4425 30389 I wm_task_moved: [530,1,6]
02-19 11:20:40.080  4425 30389 I am_create_task: [0,531]
02-19 11:20:40.080  4425 30389 I am_create_activity: [0,23723165,531,org.measurementlab.apps.Measure/org.chromium.BackgroundActivity,NULL,NULL,NULL,276889604]
02-19 11:20:40.081  4425 30389 I wm_task_created: [531,1]
02-19 11:20:40.081  4425 30389 I wm_task_moved: [531,1,7]
02-19 11:20:40.093  4425 30389 I am_pause_activity: [0,221533601,com.Slack/.ui.HomeActivity]
02-19 11:20:40.094  4425 30389 I am_home_stack_moved: [0,0,1,1,startedActivity setFocusedActivity]
02-19 11:20:40.095  4425 30389 I wm_task_moved: [531,1,7]
02-19 11:20:40.101  4425 30389 I am_focused_activity: [0,org.measurementlab.apps.Measure/org.chromium.BackgroundActivity]
02-19 11:20:40.146 15672 15672 I am_on_paused_called: [0,com.Slack.ui.HomeActivity]
02-19 11:20:40.149  4425 30388 I am_restart_activity: [0,23723165,531,org.measurementlab.apps.Measure/org.chromium.BackgroundActivity]
02-19 11:20:40.220  4425 24678 I am_home_stack_moved: [0,0,1,1,sourceStackToFront]
02-19 11:20:40.220  4425 24678 I wm_task_moved: [531,1,7]
02-19 11:20:40.223  4425 24678 I am_create_activity: [0,74368409,531,org.measurementlab.apps.Measure/.MainActivity,NULL,NULL,NULL,65540]
02-19 11:20:40.224  4425 24678 I am_pause_activity: [0,23723165,org.measurementlab.apps.Measure/org.chromium.BackgroundActivity]
02-19 11:20:40.225  4425 24678 I am_home_stack_moved: [0,0,1,1,startedActivity setFocusedActivity]
02-19 11:20:40.225  4425 24678 I wm_task_moved: [531,1,7]
02-19 11:20:40.227  4425 24678 I am_focused_activity: [0,org.measurementlab.apps.Measure/.MainActivity]
02-19 11:20:40.302  4425  5694 I am_crash: [16038,0,org.measurementlab.apps.Measure,952680006,java.lang.IllegalStateException,Activity {org.measurementlab.apps.Measure/org.chromium.BackgroundActivity} did not call finish() prior to onResume() completing,Activity.java,6339]
02-19 11:20:40.726  4425  4445 I am_restart_activity: [0,74368409,531,org.measurementlab.apps.Measure/.MainActivity]
02-19 11:20:40.787  4425  4444 I am_pss  : [8342,10069,com.google.android.gm,59985920,56696832]
02-19 11:20:41.284  4425  4445 I am_proc_start: [0,16073,10114,com.foursquare.robin,broadcast,com.foursquare.robin/com.foursquare.pilgrim.service.ReceiverPilgrimLocationClientFire]
02-19 11:20:41.337  4425 30389 I am_proc_bound: [0,16073,com.foursquare.robin]
02-19 11:20:41.580  4425 30389 I am_proc_died: [0,16038,org.measurementlab.apps.Measure]
02-19 11:20:41.581  4425 30389 I am_focused_activity: [-1,NULL]
02-19 11:20:41.699  4425 30389 I am_proc_start: [0,16090,10187,org.measurementlab.apps.Measure,activity,org.measurementlab.apps.Measure/.MainActivity]
02-19 11:20:41.743  4425  7104 I am_proc_bound: [0,16090,org.measurementlab.apps.Measure]
02-19 11:20:41.746  4425  7104 I am_restart_activity: [0,74368409,531,org.measurementlab.apps.Measure/.MainActivity]
02-19 11:20:41.768   371   371 I sf_frame_dur: [Application Error: org.measurementlab.apps.Measure,4,1,0,1,0,1,0]
02-19 11:20:42.000   371   371 I sf_frame_dur: [com.Slack/com.Slack.ui.HomeActivity,27,2,2,2,1,1,0]
02-19 11:20:42.341 16090 16090 I am_on_resume_called: [0,org.measurementlab.apps.Measure.MainActivity]
02-19 11:20:42.643  4425  4461 I am_activity_launch_time: [0,74368409,org.measurementlab.apps.Measure/.MainActivity,1919,2497]
02-19 11:20:45.704  4425  6187 I am_proc_died: [0,15576,com.instagram.android]
02-19 11:20:46.676  4425  4445 I am_proc_start: [0,16177,10170,com.bitstrips.imoji,service,com.bitstrips.imoji/.services.OutfitSaverService]
02-19 11:20:46.731  4425  5694 I am_proc_bound: [0,16177,com.bitstrips.imoji]
02-19 11:20:46.824  4425  4444 I am_pss  : [15765,10150,com.facebook.katana:videoplayer,12951552,11501568]
02-19 11:20:50.145  4425  6187 I am_proc_died: [0,15826,com.facebook.moments]
02-19 11:20:50.430  4425  4438 I am_home_stack_moved: [0,1,0,0,bringingFoundTaskToFront setFocusedActivity]
02-19 11:20:50.430  4425  4438 I wm_task_moved: [288,1,1]
02-19 11:20:50.433  4425  4438 I am_focused_activity: [0,com.android.systemui/.recents.RecentsActivity]
02-19 11:20:50.446  4425  4438 I am_pause_activity: [0,74368409,org.measurementlab.apps.Measure/.MainActivity]
02-19 11:20:50.448  4425  4438 I am_resume_activity: [0,264631282,288,com.android.systemui/.recents.RecentsActivity]
02-19 11:20:50.450  4425  4438 I am_task_to_front: [0,288]
02-19 11:20:50.450  4425  4438 I am_new_intent: [0,267801145,288,com.android.systemui/.recents.RecentsActivity,com.android.systemui.recents.SHOW_RECENTS,NULL,NULL,281034752]
```

**Comment by georgiamoon** [comment]
That would describe the experience I was having, yes. I missed a bunch of slack notifications, that would make sense. Let me try the changes now.

**Comment by sajacy** [comment]
@georgiamoon @collina 

The main issue with server selection was that the user-selected value was being stored within a local variable (vs saved to storage, which is accessible between the UI page and background page).  I removed the local variable and cleaned up the access pattern.  

While debugging, I also noticed extraneous arguments and general cleanliness issues.  I'll point them out in diff comments...

File: www/js/services/clientService.js [hunk 1]
```diff
@@ -40,39 +40,30 @@ angular.module('Measure.services.MeasurementClient', [])
     };
 
     var next = Math.max(Math.min(current + (DELTAS[state] || 0), CEILINGS[state] || 1), FLOORS[state] || 0);
-    return next;
+    return Math.round(next + 'e2') * 1e-2;
```
Comment (sajacy) [inline www/js/services/clientService.js:43]: Round so we don't get floating-point errors (e.g. `0.06` vs `0.06000000000004`)

File: www/js/services/clientService.js [hunk 2]
```diff
@@ -40,39 +40,30 @@ angular.module('Measure.services.MeasurementClient', [])
...
   }
 
   var MeasurementClientService = {
-    "start": function start(server, port, path, interval, backgroundInitiated) {
```
Comment (sajacy) [inline www/js/services/clientService.js:36]: Realized these are all unused / unnecessary arguments (can be derived solely from settings and constants).

File: www/js/services/clientService.js [hunk 3]
```diff
@@ -40,39 +40,30 @@ angular.module('Measure.services.MeasurementClient', [])
...
-
-      MLabService.findServer(setMetroSelection).then(function(mlabAnswer) {
-        measurementRecord.mlabInformation = angular.copy(mlabAnswer);
-        MeasurementService.start(measurementRecord.mlabInformation.fqdn, port, path, interval).then(function(passedResults) {
+      $q.all({
+        'accessInformation': accessInformation.getAccessInformation(),
+        'mlabInformation': SettingsService.get('metroSelection').then(MLabService.findServer)
+      })
+      .then(function(info) {
+        angular.merge(measurementRecord, info);
```
Comment (sajacy) [inline www/js/services/clientService.js:63]: A nicer idiom for merging HTTP results into a local object.

File: www/js/services/clientService.js [hunk 4]
```diff
@@ -92,8 +82,7 @@ angular.module('Measure.services.MeasurementClient', [])
           var testStatus = deferredNotification.testStatus,
             passedResults = deferredNotification.passedResults;
           progress = incrementProgress(progress, testStatus);
-          ChromeAppSupport.notify("measurement:status", {'testStatus': testStatus, 'passedResults': passedResults, "running": true, "progress": progress });
-          $rootScope.$emit(emitKey, {'testStatus': testStatus, 'passedResults': passedResults});
```
Comment (sajacy) [inline www/js/services/clientService.js:85]: Removed unused `$rootScope.$emit`

File: www/templates/measure.html [hunk 5]
```diff
@@ -8,19 +8,16 @@
...
-				<i class="icon ion-earth"
-						ng-show="mlabInformation.label === undefined"></i>
-				<a ui-sref="app.serverSelection()"
-					ng-show="mlabInformation.label !== undefined"
-					class="button button-clear button-dark
-						currentLocation">{{ mlabInformation.label }}</a>
+				<i class="icon ion-earth" ng-show="!mlabInformation.label"></i>
+				<a ui-sref="app.serverSelection()" ng-show="mlabInformation.label"
+                                  class="button button-clear button-dark currentLocation"
+                                  ng-bind="mlabInformation.label"></a>
```
Comment (sajacy) [inline www/templates/measure.html:14]: Moved bindings from `{{...}}` to `ng-bind` statements.

File: www/templates/measure.html [hunk 6]
```diff
@@ -66,7 +57,7 @@
 					translate>Finding Server</span>
 			</div>
         </div>
-		<div class="row" ng-show="lastMeasurementId !== undefined && !measurementRunning">
+        <div class="row" ng-show="lastMeasurementId > -1 && !measurementRunning">
```
Comment (sajacy) [inline www/templates/measure.html:60]: Fixed a minor UI bug that left the "More +" button visible even without any measurements.

File: www/js/controllers/measureCtrl.js [hunk 7]
```diff
@@ -80,29 +80,25 @@ angular.module('Measure.controllers.Measurement', [])
...
+  var tryConnectivity = function tryConnectivity() {
+    return $q.all({
+      'accessInformation': accessInformation.getAccessInformation(),
+      'mlabInformation': SettingsService.get('metroSelection').then(MLabService.findServer)
+    }).then(function(info) {
+      angular.merge($scope, info);
       $ionicLoading.hide();
+    }, function() {
+      var pollConnectivity = $interval(function() {
+        tryConnectivity().then(function() { $interval.cancel(pollConnectivity); });
```
Comment (sajacy) [inline www/js/controllers/measureCtrl.js:92]: Added a naive 10s retry when trying to ascertain closest server and current ISP. (so someone can leave the app open and just roam around until they get signal).

File: www/js/controllers/measureCtrl.js [hunk 8]
```diff
@@ -133,23 +127,8 @@ angular.module('Measure.controllers.Measurement', [])
...
+    ChromeAppSupport.notify('measurement:start');
     $scope.currentState = 'Starting';
-
-    if ($scope.mlabInformation === undefined) {
-      intervalPromise = $interval(function () {
-        if ($scope.mlabInformation !== undefined) {
-          $interval.cancel(intervalPromise);
-          $scope.startNDT();
-        }
-      }, 100);
```
Comment (sajacy) [inline www/js/controllers/measureCtrl.js]: This (spiritual ancestor version of retry) was removed in favor of the retry logic in `tryConnectivity`.

File: www/js/controllers/measureCtrl.js [hunk 9]
```diff
@@ -133,23 +127,8 @@ angular.module('Measure.controllers.Measurement', [])
   };
 
   $scope.startNDT = function() {
-    ChromeAppSupport.notify('measurement:start', {
-      'server': $scope.mlabInformation.fqdn,
-      'port': 3001,
-      'path': '/ndt_protocol',
-      'interval': 200
```
Comment (sajacy) [inline www/js/controllers/measureCtrl.js]: Moved these constants into the clientService.js implementation.

**Pull request merged**